04-spatial-operations.Rmd

```diff
@@ -82,6 +89,7 @@ These determine the type of spatial relationship that features in the target obj
 *Intersects* is the default spatial subsetting operator, a default that returns `TRUE` for many types of spatial relations, including *touches*, *crosses* and *is within*.
 These alternative spatial operators can be specified with the `op =` argument, as illustrated below (try plotting the result or skip to section \@ref(topological-relations) to find out what this does):
 
+<!-- any reason why you use , , instead of nz_height[canterbury, , op = st_disjoint]. Additionally, I think it would be more satisfying for the reader to already state here that st_disjoint is the opposite of st_intersects  -->
```

Comment:
Any reason why you use , , instead of nz_height[canterbury, , op = st_disjoint]. Additionally, I think it would be more satisfying for the reader to already state here that st_disjoint is the opposite of st_intersects.
Pls note also a remark below where I will suggest to switch places for sections subsetting and topological relations

Reply:
Agree that it's more satisfying to *know* what it does (although it's good to provide incentives to keep reading!).

I've updated the line in any case so it now reads:

```
These alternative spatial operators can be specified with the `op =` argument.
This is demonstrated in the following command which returns the opposite of `st_intersect()`, points that do not intersect with Cantebury (see in section \@ref(topological-relations)):
```

However, I'm not sure what you mean by the following:

> any reason why you use , , instead of nz_height[canterbury, , op = st_disjoint].

I didn't use `, , ` but:

```r
nz_height[canterbury, , op = st_disjoint]
```

which is what you're recommended, unless I'm missing something or you made a typo in that comment.

Reply:
😄 , my mistake. I was wondering, why you use 
`nz_height[canterbury, , op = st_disjoint]`
instead of:
`nz_height[canterbury, op = st_disjoint]`

Reply:
To show that the `[` operator expects a third argument. Is there a strong reason to remove it? I think it risks people thinking the second argument in `[` for spatial objects is `op`, which is untrue: that is a good reason for it to stay 'as is'.

Reply:
Just looks a bit strange, and therefore might be confusing. On the other hand I like that you make clear that all columns should be selected. However this would also imply to always use a comma prior to a column name in column subsetting, e.g., `nz_height[, "height"]` instead of `nz_height["height"]` as is sometimes done in our code chunks.

Reply:
That's a good point about `obj[col]` vs 'obj[, col]`.

There is an important difference for data frames, as illustrated by:

``` r
head(cars["dist"])
#>   dist
#> 1    2
#> 2   10
#> 3    4
#> 4   22
#> 5   16
#> 6   10
head(cars[, "dist"])
#> [1]  2 10  4 22 16 10
```

This makes me think of a more important improvement: mention that `obj[col]` is the same as `obj[,col]` for `sf` objects but not `data.frame`s in c3. Up for giving that a bash?

I think mentioning that the empty `, ,` part means 'all columns' is the best solution - `obj[col]` is frequently used and widely understood but `obj[sp_ob, , op = st_*]` is not so should be spelt out explicitly to avoid the much greater risk of causing confusion.

Reply:
Sounds good. And yes, I can mention this in c3.

One advantage of **dplyr** is that it always returns a dataframe object. In base R this is not the case as shown by your `cars` example. However, `cars[, "dist", drop = FALSE]` would also return a dataframe. In the case of sf-objects a dataframe object is returned when you only select a single attribute, I guess, because of the sticky geometry column, i.e., in fact at least two columns are returned which avoids the automatic coercion into the lowest possible dimension as done by  `[`.

Reply:
Interesting points worth mentioning. A good candidate for an rmdnote?

Reply:
The explanation of the empty `, ,` bit has now been added. Tag me if/when you add a not in c3 and I'll take a look.